File: 2020/9.kt
// Advent of code 2020 day 9
import java.io.File

fun main() {
  //println(Int.MAX_VALUE)
  val fileName = "input9.txt"
  val numbers: List<Long> = File(fileName).readLines().map { it.toLong() }

  val preamble = 25

  val active = mutableSetOf<Long>()
  var i = 0
  while (i < preamble) {
    active.add(numbers[i])
    i += 1
  }

  // Part 1: find the invalid number
  var invalid: Long

  while (true) {
    val next_number: Long = numbers[i]
    var found = false
    for (number: Long in numbers.subList(i - preamble, i)) {
      val required: Long = next_number - number
      if (active.contains(required)) {
        found = true
        break
      }
    }
    if (!found) {
      invalid = next_number
      break
    }
    active.add(next_number)
    active.remove(numbers[i - preamble])
    i += 1
  }

  println("Part 1: $invalid")

  // Part 2: find contiguous numbers that add to the invalid number
  i = 0 // the start of the string of numbers

  var found = false
  var smallest = -1
  var biggest = -1
  while (!found) {
    var j = i + 1 // the index inside the string of numbers
    var total = numbers[i] + numbers[j]

    while (total <= invalid) {
      if (total == invalid) {
        println("Got it! ${ numbers.subList(i, j + 1)}")
        smallest = i
        biggest = j
        found = true
        break
      }
      j += 1
      total += numbers[j]
    }
    // if we get to here, we went over the number we're looking for
    i += 1
  }

  val contiguous = numbers.subList(smallest, biggest + 1).sorted()
  println("Part 2: ${contiguous[0] + contiguous.last()}")
}

File: 2020/9.py
#!/usr/bin/env python

inputfile = "input9.txt"
with open(inputfile, "r") as f:
  numbers = [int(x.strip()) for x in f.readlines()]

preamble = 25
"""
numbers = [
  35,
  20,
  15,
  25,
  47,
  40,
  62,
  55,
  65,
  95,
  102,
  117,
  150,
  182,
  127,
  219,
  299,
  277,
  309,
  576,
]

preamble = 5
"""

active = set()
i = 0
for i in range(preamble):
  active.add(numbers[i])
  i += 1


# Part 1: find the invalid number

invalid = -1

while True:
  next_number = numbers[i]
  found = False
  for j in range (i - preamble, i):
    required = next_number - numbers[j]
    if required in active:
      found = True
      break
  if not found:
    invalid = next_number
    break
  active.add(next_number)
  # TODO this can't handle duplicates?
  active.remove(numbers[i - preamble])
  i += 1

print("%d didn't work" % invalid)

# Part 2: find contiguous numbers that add to the invalid number
i = 0 # the start of the string of numbers

found = False
smallest = -1
biggest = -1
while not found:
  print "i is", i
  j = i + 1 # the index inside the string of numbers
  total = numbers[i] + numbers[j]

  while total <= invalid:
    print "j is", j, "(", numbers[j], ")"
    if total == invalid:
      print("Got it! %s\n" % numbers[i: j + 1])
      smallest = i
      biggest = j
      found = True
      break
    j += 1
    total += numbers[j]
  # if we get to here, we went over the number we're looking for
  i += 1

contiguous = sorted(list(numbers[smallest: biggest + 1]))
print contiguous[0] + contiguous[-1]
